Controllers/LoginController.cs

```diff
@@ -1,54 +1,85 @@
 ﻿using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Authorization;
 using Practice.ViewModels;
 using Practice.Utilities;
+using System.Security.Claims;
 
 namespace Practice.Controllers
 {
     public class LoginController : BaseUserCredentialsController
     {
-        public IActionResult Index(bool loginRequired = false)
+        [AllowAnonymous]
+        public IActionResult Index()
         {
-            if (loginRequired)
+            return RedirectToHomeIfAuthenticated(() =>
             {
                 var model = new LoginViewModel
                 {
-                    PageError = "You must login first"
+                    PageError = (string?)TempData["Error"],
+                    SuccessMessage = (string?)TempData["SuccessMessage"]
                 };
                 return View(model);
-            }
-            return View();
+            });
         }
 
-        public IActionResult Authenticate(LoginViewModel model)
+        [AllowAnonymous]
+        public async Task<IActionResult> Authenticate(LoginViewModel model)
         {
-            if (model.User == null)
+            return await RedirectToHomeIfAuthenticated(async () =>
             {
-                model.PageError = "User not set, something went wrong";
-                return View(nameof(Index), model);
-            }
-            if (!ModelState.IsValid)
-            {
-                model.UsernameError = ModelState[_usernameErrorKey]?.Errors.FirstOrDefault()?.ErrorMessage;
-                model.PasswordError = ModelState[_passwordErrorKey]?.Errors.FirstOrDefault()?.ErrorMessage;
-                return View(nameof(Index), model);
-            }
-            
+                if (model.User == null)
+                {
+                    model.PageError = "User not set, something went wrong";
+                    return View(nameof(Index), model);
+                }
+                if (!ModelState.IsValid)
+                {
+                    model.UsernameError = ModelState[_usernameErrorKey]?.Errors.FirstOrDefault()?.ErrorMessage;
+                    model.PasswordError = ModelState[_passwordErrorKey]?.Errors.FirstOrDefault()?.ErrorMessage;
+                    return View(nameof(Index), model);
+                }
 
-            var existing = UserService.FindByUsername(model.User.Username);
-            var hashedPwd = Encrypter.EncryptSHA256(model.User.Password);
-            if (existing == null || existing.Password != hashedPwd)
-            {
-                model.PageError = "Username or password is not correct";
-                return View(nameof(Index), model);
-            }
 
-            HttpContext.Session.SetString(_sessionKeyName, model.User.Username);
-            return RedirectToAction("Index", "Home");
+                var existing = UserService.FindByUsername(model.User.Username);
```

Comment:
We should define `FindByUserLoginInfo(string userName, string password)`, then move the logic below to this method.